### PR TITLE
Constrained tin package (Implementation of #1526)

### PR DESCRIPTION
**Pull request by kochizufan**
Please fill in this template.

- [*] Use a meaningful title for the pull request. Include the name of the package modified.
- [*] Have read [How To Contribute](https://github.com/Turfjs/turf/blob/master/CONTRIBUTING.md#how-to-contribute).
- [*] Run `npm test` at the sub modules where changes have occurred.
- [*] Run `npm run lint` to ensure code style at the turf module level. => Result: missing script lint

Submitting a new TurfJS Module.

- [*] Overview description of proposed module.
- [*] Include JSDocs with a basic example.
- [*] Execute `./scripts/generate-readmes` to create `README.md`.
- [*] Add yourself to **contributors** in `package.json` using "Full Name <@GitHub Username>".

I implemented #1526 using https://github.com/mikolalysenko/cdt2d .
Sorry my lack of knowledge of typescript, I load "cdt2d" by `require('cdt2d');`, not by `import`.

### CODE REVIEW

File: packages/turf-constrained-tin/test.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+const fs = require('fs');
+const path = require('path');
+const test = require('tape');
+const tin = require('./').default;
+
+const points = require(path.join(__dirname,  'test', 'Points.json'));
```
Comment (stebogit) [inline packages/turf-constrained-tin/test.js]: I'd probably add few more tests, for example with some edge cases to can think of

File: packages/turf-constrained-tin/index.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            z = p.properties[options.z];
+        } else if (p.geometry.coordinates.length === 3) {
+            isPointZ = true;
+            z = "" + p.geometry.coordinates[2];
+        }
+        prev[0].push(xy);
+        prev[1].push(z);
+        return prev;
+    }, [[] as any, [] as any]);
+    delete options.z;
```
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/index.ts]: This "options" object is directly passed to cdt2d library.
"delaunay""interior""exterior""infinity" those 4 attributes are cdt2d's attributes, only "z" attribute is added for this implementation.
I don't know to keep this "z" causes side effects or not, but to remove any concerns, I delete "z" from options.

File: packages/turf-constrained-tin/index.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+ * triangle.
+ *
+ * @name constrainedTin
+ * @param {FeatureCollection<Point>} points input points
+ * @param {Array<Array<number>>} [edges] list of edges
+ * @param {Object} [options] option switches
+ *   {boolean} [delaunay]: if this flag is set to true, then the resulting triangulation is converted to a Delaunay triangulation by edge flipping.
+ *   Otherwise if it is false, then an arbitrary triangulation is returned. (Default true)
+ *   {boolean} [interior]: if set, only return interior faces. See note. (Default true)
+ *   {boolean} [exterior]: if set, only return exterior faces. See note. (Default true)
```
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/index.ts]: Those options are coming from cdt2d libraries, so I copy & paste the descriptions from cdt2d's document.
But if they cause misunderstanding, how about this?

[interior]: (Default true) if it is false, interior faces are not output.
[exterior]: (Default true) if it is false, exterior faces are not output.

File: packages/turf-constrained-tin/bench.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+const fs = require('fs');
+const path = require('path');
+const Benchmark = require('benchmark');
+const tin = require('./').default;
+
+const points = JSON.parse(fs.readFileSync(path.join(__dirname, 'test', 'Points.json')));
+
+const suite = new Benchmark.Suite('turf-tin');
```
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/bench.js]: Do you mean to add the resulting output of this script?
I made some modifications and just run `node bench.js`,
it shows just 
```
turf-constrained-tin: 
``` 
Is this OK?

File: packages/turf-constrained-tin/LICENSE [hunk 5]
```diff
@@ -0,0 +1,20 @@
+The MIT License (MIT)
+
+Copyright (c) 2017 TurfJS
```
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/LICENSE]: OK I will fix

File: packages/turf-constrained-tin/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+ * var tin = turf.constrainedTin(points, [], {z: 'z'});
+ *
+ * //addToMap
+ * var addToMap = [tin, points]
+ * for (var i = 0; i < tin.features.length; i++) {
+ *   var properties  = tin.features[i].properties;
+ *   properties.fill = '#' + properties.a + properties.b + properties.c;
+ * }
+ */
+export default function constrainedTin(points: FeatureCollection<Point, any>, edges?: number[][], options?: {
```
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/index.d.ts]: I try to change ts to js but I gave up...
index.ts => index.js is easy, but I can't set up a building environment (correct package.json, etc) for javascript...
Keeping typescript is OK?

File: packages/turf-constrained-tin/LICENSE (hunk 5 above)
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/LICENSE]: Fixed

File: packages/turf-constrained-tin/index.d.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/index.d.ts]: Not fixed, reason is written above

File: packages/turf-constrained-tin/bench.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/bench.js]: Add output

File: packages/turf-constrained-tin/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/index.ts]: Fixed as written above

File: packages/turf-constrained-tin/test.js (hunk 1 above)
Comment (kochizufan) [reply packages/turf-constrained-tin/test.js]: Add 10 more tests including error case

File: packages/turf-constrained-tin/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/index.ts]: Not fixed, reason is written above

**Comment by dpmcmlxxvi** [comment]
@kochizufan @stebogit  There was an effort to implement constrained tin before #241. But it seemed to be dropped. I couldn't track down why. Maybe it was an implementation issue, maybe the developer dropped off, or maybe it was a project decision that there wasn't enough demand to include it.

Regardless, you may want to compare this with other triangulation libraries to see how it stacks up. See [here](https://github.com/mapbox/earcut#why-another-triangulation-library) for a throughput comparison of some libraries. Unfortunately, cdt2d is not included in that list. It might hep to contact the earcut developers and see if they published their test bed somewhere and you can drop this in to compare.

File: packages/turf-constrained-tin/LICENSE (hunk 5 above)
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/LICENSE]: Change year value in LICENSE to 2019 causes CI error.
Revert to 2017

File: packages/turf-constrained-tin/bench.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (stebogit) [reply packages/turf-constrained-tin/bench.js]: @kochizufan I meant the actual output of the benchmark result. Look for example at the `bench.js` of some other turf package

File: packages/turf-constrained-tin/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (stebogit) [reply packages/turf-constrained-tin/index.ts]: @kochizufan I'm not sure I understand the reason you're referring to.
My point is that a turf module should never unexpectedly change any of its inputs. Thus if I pass an `options` object to the `constrainedTin` function I should not see it changed after calling that function.

**Comment by stebogit** [comment]
Good catch @dpmcmlxxvi

File: packages/turf-constrained-tin/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/index.ts]: OK understood
Then I make Object.assign to objects and give it to cdt2d.

[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/index.ts]: Done

File: packages/turf-constrained-tin/bench.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/bench.js]: Understood.
I added some benchmark results.

**Comment by kochizufan** [comment]
@dpmcmlxxvi @stebogit I did benchmark comparison with turf-tin in bench.js
Results are:
 * constrained Tin 1 edge: 12.816ms
 * constrained Tin 2 edges: 1.956ms
 * constrained Tin 3 edges: 1.227ms
 * normal Tin (as reference): 0.812ms
 * constrained Tin 1 edge x 1,720 ops/sec ±0.35% (96 runs sampled)
 * constrained Tin 2 edges x 1,715 ops/sec ±0.52% (95 runs sampled)
 * constrained Tin 3 edges x 1,701 ops/sec ±0.63% (95 runs sampled)
 * normal Tin (as reference) x 28,858 ops/sec ±2.21% (92 runs sampled)

Unfortunately, normal (no constrained) tin is more than 10 times faster than constrained tin.
But I believe, "constrained" itself has value.
People who don't need constrained should use turf-tin, and who need constrained should use turf-constrained-tin.

**Comment by kochizufan** [comment]
I found another constrained tin library 
https://github.com/savithru-j/cdt-js
I will try this also, if it is faster than cdt2d, I will change logic into cdt-js

**Comment by kochizufan** [comment]
Benchmark result:

constrained Tin (cdt2d) 1 edge: 16.934ms
new constrained Tin (cdt-js) 1 edge: 3.424ms
constrained Tin (cdt2d) 2 edges: 3.220ms
new constrained Tin (cdt-js) 2 edges: 0.799ms
constrained Tin (cdt2d) 3 edges: 1.220ms
new constrained Tin (cdt-js) 3 edges: 0.507ms
normal Tin (as reference): 1.204ms
constrained Tin (cdt2d) 1 edge x 955 ops/sec ±1.88% (83 runs sampled)
new constrained Tin (cdt-js) 1 edge x 31,997 ops/sec ±4.52% (82 runs sampled)
constrained Tin (cdt2d) 2 edges x 995 ops/sec ±1.30% (83 runs sampled)
new constrained Tin (cdt-js) 2 edges x 33,642 ops/sec ±1.34% (83 runs sampled)
constrained Tin (cdt2d) 3 edges x 1,003 ops/sec ±1.42% (84 runs sampled)
new constrained Tin (cdt-js) 3 edges x 33,418 ops/sec ±1.32% (83 runs sampled)
normal Tin (as reference) x 17,226 ops/sec ±6.46% (82 runs sampled)

cdt-js is far faster than cdt2d, and it even somehow faster than legacy turf-tin.
So I will change my logic to cdt-js.

**Comment by kochizufan** [comment]
Changing base logic from cdt2d to cdt-js has done.

File: yarn.lock [hunk 7]
```diff
@@ -159,6 +159,13 @@
   dependencies:
     "@turf/helpers" "^5.1.5"
 
+"@turf/kinks@5.1.x", "@turf/kinks@^5.1.5":
+  version "5.1.5"
```
Comment (stebogit) [inline yarn.lock]: Is there any reason why this global file should be change?

File: packages/turf-constrained-tin/index.js [hunk 8]
```diff
@@ -0,0 +1,897 @@
...
+ * //addToMap
+ * var addToMap = [tin, points]
+ * for (var i = 0; i < tin.features.length; i++) {
+ *   var properties  = tin.features[i].properties;
+ *   properties.fill = '#' + properties.a + properties.b + properties.c;
+ * }
+ */
+module.exports = function(points, edges, z) {
+    var isPointZ = false;
+    var ret = {
```
Comment (stebogit) [inline packages/turf-constrained-tin/index.js]: Maybe you could find a more descriptive name for this `ret` variable?

File: packages/turf-constrained-tin/index.js [hunk 9]
```diff
@@ -0,0 +1,897 @@
...
+}
+
+function getPointOrientation(edge, p)
+{
+    var vec_edge01 = edge[1].sub(edge[0]);
+    var vec_edge0_to_p = p.sub(edge[0]);
+    return cross(vec_edge01, vec_edge0_to_p);
+}
+
+//Some variables for rendering
```
Comment (stebogit) [inline packages/turf-constrained-tin/index.js]: Maybe this is part of the library, but I'm not sure I understand what these apparently arbitrary values represent. Particularly the screenL(ength?) 🤔

File: packages/turf-constrained-tin/bench.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+const Benchmark = require('benchmark');
+const tin = require('@turf/tin').default;
+const constrainedTin = require('./');
+
+const points = JSON.parse(fs.readFileSync(path.join(__dirname, 'test', 'Points.json')));
+
+/**
+ * Benchmark Results
+ *
+ * constrained Tin (cdt-js) with 1 edge: 5.494ms
```
Comment (stebogit) [inline packages/turf-constrained-tin/bench.js]: This result is soooo slower compares to the others. Is it correct?

File: packages/turf-constrained-tin/test.js [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+const path = require('path');
+const test = require('tape');
+const constrainedTin = require('./');
+
+const points = require(path.join(__dirname,  'test', 'Points.json'));
+const points_grid = require(path.join(__dirname,  'test', 'Points_grid.json'));
+const mercator = require(path.join(__dirname,  'test', 'mercator.json'));
+
+test('tin - z property', t => {
+    const expected = require(path.join(__dirname,  'test', 'Tin.json'));
```
Comment (stebogit) [inline packages/turf-constrained-tin/test.js]: Sorry, I only noticed this now. Each turf package uses a similar pattern here, which is dividing test input and expected results into `in` and `out` folders (see [here](https://github.com/Turfjs/turf/tree/master/packages/turf-circle/test) for example). This makes the test code [simpler](https://github.com/Turfjs/turf/blob/master/packages/turf-circle/test.js#L25) and easier to add new test cases if necessary.

File: packages/turf-constrained-tin/test.js [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    const tinned = constrainedTin(points, [[2, 12]], 'elevation');
+    t.equal(tinned.features[0].geometry.type, 'Polygon');
+    t.equal(tinned.features.length, 24);
+    t.deepEqual(tinned, expected, 'tinned polygons match');
```
Comment (stebogit) [inline packages/turf-constrained-tin/test.js]: If you are comparing the result with a specific output you don't really need to check if the result's geometry and length, that's implicit in the deep comparison, isn't it?

File: packages/turf-constrained-tin/index.js [hunk 13]
```diff
@@ -0,0 +1,897 @@
...
+ * for developing elevation contour maps or stepped heat visualizations.
+ *
+ * If an optional z-value property is provided then it is added as properties called `a`, `b`,
+ * and `c` representing its value at each of the points that represent the corners of the
+ * triangle.
+ *
+ * @name constrainedTin
+ * @param {FeatureCollection<Point>} points input points
+ * @param {Array<Array<number>>} [edges] list of edges
+ * @param {String} [z] name of the property from which to pull z values
```
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/index.js]: I decided this argument's structure as follows:

About [edges]: Edges is optional, but, if it is not given, this module provides the same function with legacy turf-tin. 
So, it is expected almost all people who use this module won't omit this argument. That is why I made this main argument.

About [z]: Behaviors of this module are referring to legacy turf-tin module.
In turf-tin module, "z" is in the main argument, not in options object.

Those are why I made them as main arguments.

File: packages/turf-line-arc/index.js [hunk 14]
```diff
@@ -1,10 +1,7 @@
 "use strict";
-var __importDefault = (this && this.__importDefault) || function (mod) {
-    return (mod && mod.__esModule) ? mod : { "default": mod };
-};
 Object.defineProperty(exports, "__esModule", { value: true });
-var circle_1 = __importDefault(require("@turf/circle"));
```
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-line-arc/index.js]: Sorry It is not intentional.
Maybe I did it by mistake.
I will revert it in next commit.

File: yarn.lock (hunk 7 above)
Comment (kochizufan) [reply yarn.lock]: Sorry It is not intentional.
Maybe I did it by mistake.
I will revert it in next commit.

File: packages/turf-constrained-tin/test.js (hunk 11 above)
Comment (kochizufan) [reply packages/turf-constrained-tin/test.js]: OK, I will divide 'in' and 'out'.

File: packages/turf-constrained-tin/bench.js (hunk 10 above)
Comment (kochizufan) [reply packages/turf-constrained-tin/bench.js]: I don't know the reason, but it is the first time of calling the function in this benchmark, so I think maybe there are some initializations...

I add dummy function calls of both turf-constrained-tin and turf-tin before testing, it becomes better.

constrained Tin (cdt-js) with 1 edge: 1.398ms
constrained Tin (cdt-js) with 2 edges: 0.973ms
constrained Tin (cdt-js) with 3 edges: 1.915ms
normal Tin (as reference, no constrained): 0.359ms
constrained Tin (cdt-js) with 1 edge x 21,375 ops/sec ±3.26% (80 runs sampled)
constrained Tin (cdt-js) with 2 edges x 23,076 ops/sec ±4.29% (75 runs sampled)
constrained Tin (cdt-js) with 3 edges x 23,855 ops/sec ±1.70% (81 runs sampled)
normal Tin (as reference, no constrained) x 21,021 ops/sec ±1.56% (83 runs sampled)

File: packages/turf-constrained-tin/test.js (hunk 12 above)
Comment (kochizufan) [reply packages/turf-constrained-tin/test.js]: OK you are right.

File: packages/turf-constrained-tin/index.js (hunk 9 above)
Comment (kochizufan) [reply packages/turf-constrained-tin/index.js]: It is coming from the original implementation.
https://github.com/savithru-j/cdt-js/blob/master/cdt.js

The original implementation did not separate the UI from the computation logic.
So I made many efforts to remove UI-related variables and functions but only 
* min_coord
* screenL
* boundingL
these 3 variables are used in both UI and computation logic.
So they are kept with their original names.

Rough my understanding is,
* min_coord - coordinate of the origin point of the field which all points must be on it.
* screenL - Width/height of the field.
* boundingL - Far away point x/y values used for the initial triangle.

So, if it should be, I will change these variable names as follow:
* min_coord -> field_origin
* screenL -> fieldSize
* boundingL -> boundingTriangleSize

File: packages/turf-constrained-tin/index.js (hunk 8 above)
Comment (kochizufan) [reply packages/turf-constrained-tin/index.js]: I will change it to "meshData"

File: packages/turf-constrained-tin/index.js [hunk 15]
```diff
@@ -0,0 +1,897 @@
+// A fast algorithm for generating constrained delaunay triangulations
+// https://www.sciencedirect.com/science/article/pii/004579499390239A
+// A robust efficient algorithm for point location in triangulations
+// https://www.cl.cam.ac.uk/techreports/UCAM-CL-TR-728.pdf
+// https://savithru-j.github.io/cdt-js/
+// Copyright 2018 Savithru Jayasinghe
```
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/index.js]: OK I will divide them to separate files

File: packages/turf-constrained-tin/index.js (hunk 9 above)
Comment (kochizufan) [reply packages/turf-constrained-tin/index.js]: Done

File: packages/turf-constrained-tin/index.js (hunk 8 above)
Comment (kochizufan) [reply packages/turf-constrained-tin/index.js]: Done

File: packages/turf-constrained-tin/index.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/index.js]: Done

File: packages/turf-constrained-tin/index.js [hunk 16]
```diff
@@ -0,0 +1,897 @@
...
+                }
+            }
+            return coord; 
+        })
+        coords[3] = coords[0];
+        return helper.polygon([coords], properties);
+    }));
+};
+
+var Point = function(x, y) {
```
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/index.js]: cdt-js is not in npm.
I agree to separate these implementations into different file

File: packages/turf-constrained-tin/test.js (hunk 11 above)
Comment (kochizufan) [reply packages/turf-constrained-tin/test.js]: Done

File: packages/turf-constrained-tin/test.js (hunk 12 above)
Comment (kochizufan) [reply packages/turf-constrained-tin/test.js]: Removed

File: packages/turf-constrained-tin/test.js [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+
+test('tin - grid with 2 edges', t => {
+    const expected = require(path.join(__dirname,  'test', 'out', 'Tin-Grid_3.json'));
+    const tinned = constrainedTin(points_grid, [[0, 5], [5, 6]]);
+    t.deepEqual(tinned, expected, 'tinned polygons match');
+    //fs.writeFileSync(path.join(__dirname, 'test', 'out', 'Tin-Grid_3_tmp.json'), JSON.stringify(tinned, null, 2));
+    t.end();
+});
+
+test('tin - error case', t => {
```
Comment (stebogit) [inline packages/turf-constrained-tin/test.js]: You could probably add some type check for the inputs here (like `points` as array or `edges` as string), making sure the resulting error message is clear

**Comment by kochizufan** [comment]
Hi @stebogit

In my opinion, I agree with the direction that one of the modules should be eliminated in the future, but I do not think that the timing is right now.

Reason is,

* turf-tin and turf-constrained-tin without edges are same functions, but they don't provide same results. They provide the answer of several possible answers. So, there is a chance that the results turf-tin and turf-constrained-tin provided are different, it becomes backward-incompatibility and breaking change.

* turf-tin is a mature module, but turf-constrained-tin is not mature. I don't find any bugs now, but it is possible that it includes unknown bugs. so I believe it is better to keep turf-tin for a while.

So, I think it would be better to keep both for a certain period of time, with a notice of future integration.

File: packages/turf-constrained-tin/test.js [hunk 18]
```diff
@@ -0,0 +1,72 @@
+const fs = require('fs');
+const path = require('path');
+const test = require('tape');
+const constrainedTin = require('./');
+
+const points = require(path.join(__dirname,  'test', 'in', 'Points.json'));
+const points_grid = require(path.join(__dirname,  'test', 'in', 'Points_grid.json'));
+const mercator1 = require(path.join(__dirname,  'test', 'in', 'mercator_1.json'));
```
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/test.js]: OK I got it

File: packages/turf-constrained-tin/test.js (hunk 17 above)
Comment (kochizufan) [reply packages/turf-constrained-tin/test.js]: Sorry I want to confirm your intention in other words...
What you recommended is, adding some gating logic to inputs, and if the type of inputs is incorrect, I should raise errors?
If so, I'll do.

If your intention is not as above, sorry for my poorness in English, would you please explain in other words?

File: packages/turf-constrained-tin/test.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (kochizufan) [reply packages/turf-constrained-tin/test.js]: Done

File: packages/turf-constrained-tin/test.js (hunk 17 above)
Comment (kochizufan) [reply packages/turf-constrained-tin/test.js]: Done

**Comment by kochizufan** [comment]
I've already used this feature in my project as an external implementation of turf.js, so I'll withdraw this pull request.
